File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/MIMETypes.java
package it.unibo.coordination.linda.presentation;

import java.util.stream.Stream;

public enum MIMETypes {

    APPLICATION_JSON("application", "json"),

    APPLICATION_YAML("application", "yaml"),

    APPLICATION_XML("application", "xml"),

    APPLICATION_PROLOG("application", "prolog"),

    TEXT_HTML("text", "html"),

    TEXT_PLAIN("text", "plain"),

    ANY("*", "*"),

    APPLICATION_ANY("application", "*");

    private final String type, subtype;

    MIMETypes(String type, String subtype) {
        this.type = type;
        this.subtype = subtype;
    }

    public String getType() {
        return type;
    }

    public String getSubtype() {
        return subtype;
    }

    @Override
    public String toString() {
        return type + "/" + subtype;
    }

    public boolean matches(String other) {
        return match(this, other);
    }

    public static boolean match(MIMETypes mime, String other) {
        if (other == null || !other.contains("/")) return false;

        final String[] parts = other.split("/");

        if (parts.length != 2) return false;

        return match(mime, parts[0], parts[1]);
    }

    private static boolean match(MIMETypes mime, String type, String subtype) {
        return (Stream.of(mime.getType(), type).anyMatch("*"::equals)
                || mime.getType().equalsIgnoreCase(type))
                && (Stream.of(mime.getSubtype(), subtype).anyMatch("*"::equals)
                        || mime.getSubtype().equalsIgnoreCase(subtype));
    }

    public static MIMETypes parse(String string) {
        return Stream.of(values()).filter(it -> it.toString().equalsIgnoreCase(string)).findAny().orElseGet(() -> {
           throw new IllegalArgumentException(string);
        });
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/MIMETypes.kt
package it.unibo.presentation

import java.util.*

enum class MIMETypes(val type: String, val subtype: String) {
    APPLICATION_JSON("application", "json"),
    APPLICATION_YAML("application", "yaml"),
    APPLICATION_XML("application", "xml"),
    APPLICATION_PROLOG("application", "prolog"),
    TEXT_HTML("text", "html"),
    TEXT_PLAIN("text", "plain"), ANY("*", "*"),
    APPLICATION_ANY("application", "*");

    override fun toString(): String {
        return "$type/$subtype"
    }

    fun matches(other: String?): Boolean {
        return match(this, other)
    }

    companion object {

        val XML_JSON_YAML = EnumSet.of(APPLICATION_XML, APPLICATION_JSON, APPLICATION_YAML)

        @JvmStatic
        fun match(mime: MIMETypes, other: String?): Boolean {
            if (other == null || !other.contains("/")) return false
            val parts = other.split("/").toTypedArray()
            return if (parts.size != 2) false else match(mime, parts[0], parts[1])
        }

        @JvmStatic
        fun match(mime: MIMETypes, type: String, subtype: String): Boolean {
            return ((sequenceOf(mime.type, type).any { "*" == it } || mime.type.equals(type, ignoreCase = true))
                    && (sequenceOf(mime.subtype, subtype).any { "*" == it }  || mime.subtype.equals(subtype, ignoreCase = true)))
        }

        @JvmStatic
        fun parse(string: String?): MIMETypes {
            return sequenceOf(*values())
                    .find { it.toString().equals(string, ignoreCase = true) }
                    ?: throw IllegalArgumentException(string)
        }
    }

}